cmd/erasure-healfile.go

```diff
@@ -36,48 +50,86 @@ func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string,
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for ; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
+		var buffer []byte
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			if disk == OfflineDisk {
+				continue
+			}
+			// allocate buffer only when needed - when
+			// reads fail, the buffer can be reused
+			if int64(len(buffer)) != chunksize {
+				buffer = make([]byte, chunksize)
+			}
+			if !verifiers[i].IsVerified() {
+				_, err = disk.ReadFileWithVerify(volume, path,
+					chunkOffset, buffer, verifiers[i])
+			} else {
+				_, err = disk.ReadFile(volume, path,
+					chunkOffset, buffer)
+			}
+			if err != nil {
+				// LOG FIXME: add a conditional log
+				// for read failures, once per-disk
+				// per-function-invocation.
+				continue
+			}
+
+			// read was successful, so set the buffer as
+			// blocks[i], and reset buffer to nil to force
+			// allocation on next iteration
+			blocks[i], buffer = buffer, nil
+
+			numReads++
+			if numReads == s.dataBlocks {
+				// we have enough data to reconstruct
+				break
 			}
 		}
+
+		// advance the chunk offset to prepare for next loop
+		// iteration
+		chunkOffset += chunksize
+
+		// reconstruct data - this computes all data and parity shards
 		if err = s.ErasureDecodeDataAndParityBlocks(blocks); err != nil {
 			return f, err
 		}
+
+		// write computed shards as chunks on file in each
+		// stale disk
 		for i, disk := range s.disks {
 			if disk != OfflineDisk {
 				continue
 			}
-			if err = offlineDisks[i].AppendFile(healVolume, healPath, blocks[i]); err != nil {
+
+			err = staleDisks[i].AppendFile(healVolume, healPath,
+				blocks[i])
+			if err != nil {
 				return f, traceError(err)
 			}
 			hashers[i].Write(blocks[i])
 		}
 	}
+
+	// copy computed file hashes into output variable
 	f.Size = size
 	f.Algorithm = algorithm
 	for i, disk := range s.disks {
```

Comment:
`for i, disk := range staleDisks {` here as well

Reply:
Fixed.